In [22]:
import pandas as pd
import numpy as np
import matplotlib as plt
import plotly.express as px

How is the data science industry changing over time?


WHAT ARE THE DEMOGRAPHICS OF THE DATA SCIENCE INDUSTRY AND HOW HAS THAT CHANGED OVER TIME? 2018-(Q1, Q2, Q3, Q4, Q9, Q5)
1.) Gender/Pay breakdown
2.) Education/Age breakdown 
3.) Undergraduate degree
3.) Country distribution over time how is that changing


HOW ARE DATA SCIENTISTS BECOMING DATA SCIENTISTS? 2018-(Q6, Q8, Q18, Q24, Q26, Q35, Q36, Q37, Q38, Q39, Q40)
1.) 

WHAT TOOLS/PROGRAMMING LANGUAGES/ENVIRONMENTS DO DATA SCIENTISTS USE AND HOW HAS THAT CHANGED OVER TIME? 2018-(Q12, Q13, Q14, Q15, Q16, Q17, Q19, Q20, Q21, Q22, Q49, Q27, Q28, Q29, Q30, Q31, Q32, Q33, Q49, Q50)
6.) Most used Programming Language? What languages are recommended to learn?
7.) Is data science happening more often in the cloud or locally? What type of IDEs?
3.) Tools they're using

WHAT TYPE OF DATA ARE DATA SCIENTISTS DEALING WITH? 
1.) Size of data
2.) Frequency of data

WHAT DO DATA SCIENTISTS DO ON A DAILY BASIS? The Role: 2018-(Q11, Q47, Q48, Q50, Q51, Q23, Q34, Q47)
1.) % breakdown of modeling, cleaning, wrangling, communicating, etc.
5.) Activities that are part of their job
2.) Major barriers to accomplishing their work
4.) 


BIAS IN DATA: WHAT ARE WE DOING TO ADDRESS IT? 2018-(Q41, Q43, Q45, Q46, Q48)
1.) Do DS/ML believe in data bias in their algorithms?
2

INDUSTRY BREAKDOWN: WHERE ARE THE MOST DATA SCIENTISTS? IS IT INDUSTRY AGNOSTIC? 2018-(Q7, Q10, Q42, Q44)

# 2019, 2018, 2017 Kaggle ML & DS Survey #
### A Comprehensive Survey Analysis of 24,000 Machine Learning & Data Science Practioners####

This year, as last year, we set out to conduct an industry-wide survey that presents a truly comprehensive view of the state of data science and machine learning. The survey was live for one week in October, and after cleaning the data we finished with 23,859 responses, a 49% increase over last year!

There's a lot to explore here. The results include raw numbers about who is working with data, what’s happening with machine learning in different industries, and the best ways for new data scientists to break into the field. We've published the data in as raw a format as possible without compromising anonymization, which makes it an unusual example of a survey dataset.

The challenge objective: tell a data story about a subset of the data science community represented in this survey, through a combination of both narrative text and data exploration. A “story” could be defined any number of ways, and that’s deliberate. The challenge is to deeply explore (through data) the impact, priorities, or concerns of a specific group of data science and machine learning practitioners. That group can be defined in the macro (for example: anyone who does most of their coding in Python) or the micro (for example: female data science students studying machine learning in masters programs). This is an opportunity to be creative and tell the story of a community you identify with or are passionate about!

Submissions will be evaluated on the following:

Composition - Is there a clear narrative thread to the story that’s articulated and supported by data? The subject should be well defined, well researched, and well supported through the use of data and visualizations.
Originality - Does the reader learn something new through this submission? Or is the reader challenged to think about something in a new way? A great entry will be informative, thought provoking, and fresh all at the same time.
Documentation - Are your code, and kernel, and additional data sources well documented so a reader can understand what you did? Are your sources clearly cited? A high quality analysis should be concise and clear at each step so the rationale is easy to follow and the process is reproducible

In [23]:
df_freeForm = pd.read_csv('kaggle-survey-2018/freeFormResponses.csv')
df_multi_choice = pd.read_csv('kaggle-survey-2018/multipleChoiceResponses.csv')
df_survey_schema = pd.read_csv('kaggle-survey-2018/SurveySchema.csv')

/Users/adrianlievano/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2728: DtypeWarning:

Columns (25) have mixed types. Specify dtype option on import or set low_memory=False.

/Users/adrianlievano/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2728: DtypeWarning:

Columns (0,2,8,10,21,23,24,25,26,27,28,44,56,64,83,85,87,107,109,123,125,150,157,172,174,194,210,218,219,223,246,249,262,264,276,277,278,279,280,281,282,283,284,285,286,287,288,289,290,304,306,325,326,329,341,368,371,384,385,389,390,391,393,394) have mixed types. Specify dtype option on import or set low_memory=False.



In [24]:
df_survey_schema.head()

2018 Kaggle Machine Learning and Data Science Survey  \
0                                          Question:     
1                                  # of Respondents:     
2  Who was excluded? (0 = not excluded; 1 = exclu...     
3  If What is your age (# years)? 0-17 Is Selecte...     
4  If What is the highest level of formal educati...     

                                       Q1  \
0  What is your gender? - Selected Choice   
1                                   23860   
2                                       0   
3                                       0   
4                                       0   

                                                 Q10  \
0  Does your current employer incorporate machine...   
1                                              20670   
2                                                  0   
3                                                  1   
4                                                  0   

                                                 Q11  \
0  Select any activities that make up an importan...   
1                                              19518   
2                                                  0   
3                                                  1   
4                                                  0   

                                                 Q12  \
0  What is the primary tool that you use at work ...   
1                                              19199   
2                                                  0   
3                                                  1   
4                                                  0   

                                                 Q13  \
0  Which of the following integrated development ...   
1                                              19117   
2                                                  0   
3                                                  1   
4                                                  0   

                                                 Q14  \
0  Which of the following hosted notebooks have y...   
1                                              18971   
2                                                  0   
3                                                  1   
4                                                  0   

                                                 Q15  \
0  Which of the following cloud computing service...   
1                                              18864   
2                                                  0   
3                                                  1   
4                                                  0   

                                                 Q16  \
0  What programming languages do you use on a reg...   
1                                              18828   
2                                                  0   
3                                                  1   
4                                                  0   

                                                 Q17  \
0  What specific programming language do you use ...   
1                                              15223   
2                                                  0   
3                                                  1   
4                                                  0   

                  ...                  \
0                 ...                   
1                 ...                   
2                 ...                   
3                 ...                   
4                 ...                   

                                                 Q47  \
0  What methods do you prefer for explaining and/...   
1                                              13418   
2                                                  0   
3                                                  1   
4                                                  0   

                                                 Q48  \
0  Do you consider ML models to be "black boxe

In [25]:
columns = df_survey_schema.columns[0:]
len(columns)

52

## Print and format all the questions in the 2018 Kaggle Survey 

In [26]:
print('There are {} questions in this survey'.format(len(df_survey_schema.iloc[0])))

for i in range(len(df_survey_schema.iloc[0])):
    
    print(' ')
    print(columns[i])
    print(df_survey_schema.iloc[0][i])
    print(' ')
    

There are 52 questions in this survey
 
2018 Kaggle Machine Learning and Data Science Survey
Question:
 
 
Q1
What is your gender? - Selected Choice
 
 
Q10
Does your current employer incorporate machine learning methods into their business?
 
 
Q11
Select any activities that make up an important part of your role at work: (Select all that apply) - Selected Choice
 
 
Q12
What is the primary tool that you use at work or school to analyze data? (include text response) - Selected Choice
 
 
Q13
Which of the following integrated development environments (IDE's) have you used at work or school in the last 5 years? (Select all that apply) - Selected Choice
 
 
Q14
Which of the following hosted notebooks have you used at work or school in the last 5 years? (Select all that apply) - Selected Choice
 
 
Q15
Which of the following cloud computing services have you used at work or school in the last 5 years? (Select all that apply) - Selected Choice
 
 
Q16
What programming languages do you use 

In [27]:
#Remove row 1 and reset index to 0 to make analysis easier
df_multi_choice = df_multi_choice[1:]
df_multi_choice.head()

Time from Start to Finish (seconds)      Q1 Q1_OTHER_TEXT     Q2  \
1                                 710  Female            -1  45-49   
2                                 434    Male            -1  30-34   
3                                 718  Female            -1  30-34   
4                                 621    Male            -1  35-39   
5                                 731    Male            -1  22-24   

                         Q3                 Q4  \
1  United States of America    Doctoral degree   
2                 Indonesia  Bachelor’s degree   
3  United States of America    Master’s degree   
4  United States of America    Master’s degree   
5                     India    Master’s degree   

                                                  Q5              Q6  \
1                                              Other      Consultant   
2                 Engineering (non-computer focused)           Other   
3      Computer science (software engineering, etc.)  Data Scientist   
4  Social sciences (anthropology, psychology, soc...    Not employed   
5                          Mathematics or statistics    Data Analyst   

  Q6_OTHER_TEXT                         Q7      ...       Q49_OTHER_TEXT  \
1            -1                      Other      ...                   -1   
2             0  Manufacturing/Fabrication      ...                   -1   
3            -1             I am a student      ...                   -1   
4            -1                        NaN      ...                   -1   
5            -1             I am a student      ...                   -1   

  Q50_Part_1          Q50_Part_2                             Q50_Part_3  \
1        NaN                 NaN                                    NaN   
2        NaN                 NaN                                    NaN   
3        NaN  Too time-consuming                                    NaN   
4        NaN                 NaN  Requires too much technical knowledge   
5        NaN  Too time-consuming                                    NaN   

  Q50_Part_4                              Q50_Part_5 Q50_Part_6 Q50_Part_7  \
1        NaN                                     NaN        NaN        NaN   
2        NaN                                     NaN        NaN        NaN   
3        NaN                                     NaN        NaN        NaN   
4        NaN  Not enough incentives to share my work        NaN        NaN   
5        NaN  Not enough incentives to share my work        NaN        NaN   

  Q50_Part_8 Q50_OTHER_TEXT  
1        NaN             -1  
2        NaN             -1  
3        NaN             -1  
4        NaN             -1  
5        NaN             -1  

[5 rows x 395 columns]

In [28]:
df_test = pd.DataFrame({'count' : df_multi_choice.groupby([ "Q1", "Q9"] ).size()}).reset_index().sort_values(by = 'count', ascending = False)
df_test.head()



Q1                                                 Q9  count
36  Male  I do not wish to disclose my approximate yearl...   3759
18  Male                                           0-10,000   3610
19  Male                                          10-20,000   1616
23  Male                                          20-30,000   1181
26  Male                                          30-40,000    933

In [29]:
fig = px.bar(df_test, x = 'Q9', y = 'count', color = 'Q1', barmode = 'group', title= 'Compensation breakdown per Gender Category')
fig.show()

In [30]:
df_test = pd.DataFrame({'count' : df_multi_choice.groupby([ "Q4"] ).size()}).reset_index().sort_values(by = 'count', ascending = False)
df_test.head()

fig = px.bar(df_test, x = 'Q4', y = 'count', title= 'Education of Respondents in 2018')
fig.show()

## Age Breakdown

In [31]:
df_test = pd.DataFrame({'count' : df_multi_choice.groupby([ "Q6", "Q2"] ).size()}).reset_index().sort_values(by = 'count', ascending = False)
df_test.head()

fig = px.bar(df_test, x = 'Q6', y = 'count', title= 'Age and Role of Survey Respondents in 2018', color = 'Q2' , barmode='group')
fig.show()

Notes:
    1. The figure above is interactive. Drag and drop a box around the role you want to learn more about. The purpose of this figure is to enable readers to explore - at a high level - the data.

In [32]:
df_test = pd.DataFrame({'count' : df_multi_choice.groupby([ "Q3", "Q1"] ).size()}).reset_index().sort_values(by = 'count', ascending = False)
df_test.head()


Q3      Q1  count
67                      India    Male   3719
177  United States of America    Male   3530
28                      China    Male   1337
176  United States of America  Female   1082
113                     Other    Male    849

In [33]:
fig = px.bar(df_test, x = 'Q3', y = 'count', color = 'Q1', barmode = 'group', title= 'Where do Survey Respondents Reside?')
fig.show()

Notes:
    1. The figure above is interactive. Drag and drop a box around the country you want to learn more about. The purpose of this figure is to enable readers to explore - at a high level - the data.

In [34]:
df_test = pd.DataFrame({'count' : df_multi_choice.groupby([ "Q3"] ).size()}).reset_index().sort_values(by = 'count', ascending = False)
df_test.head()


Q3  count
56  United States of America   4716
21                     India   4417
9                      China   1644
36                     Other   1036
44                    Russia    879

In [35]:
fig = px.bar(df_test, x = 'Q3', y = 'count', title= 'Where do Survey Respondents reside?')
fig.show()

Notes:
    1. The figure above is interactive. Drag and drop a box around the country you want to learn more about. The purpose of this figure is to enable readers to explore - at a high level - the data.

In [36]:
df_test['Q3'].unique()[:3]

array(['United States of America', 'India', 'China'], dtype=object)

In [37]:
df_test['count'][:3]

56    4716
21    4417
9     1644
Name: count, dtype: int64

In [38]:
import plotly.graph_objects as go

top_n = 10
labels = df_test['Q3'].unique()[:top_n]
values = df_test['count'][:top_n]

fig = go.Figure(data =[go.Pie(labels = labels, values = values, hole = 0.2)])
fig.update_layout(title_text = 'Top {} Countries where Survey Respondents Reside in 2018'.format(top_n))
fig.show()

#use to create 2019 and 2017 pie chart
#fig.add_trace(go.Pie(labels=labels, values=[38, 27, 18, 10, 7], name='Starry Night',
             #        marker_colors=night_colors), 1, 1)

#### Get continent column and time column indicating the year
fig = px.scatter_geo(df_test, locations = 'Q3', color = 'Q3', hover_name = 'Q3', 
                     size = 'count', projection = 'natural earth')
fig.show()

### Key takeaways from Aggregate Data:

    1. The variety of roles, countries, ages, and shows that the Kaggle Data Science community is diverse and international. The 2018 survey shows that the United States only accounts for 29.7% of survey respondents. The two largest countries in Asia (India & China) account for nearly 37.9% of all survey respondents. 
    2. 
    3. ASDA

Note: For the purposes of this analysis, we will focus on self-described students, data scientists, software engineers, and data engineers in an international context 

In [51]:
df_multi_tech = df_multi_choice[(df_multi_choice['Q6'] == 'Data Scientist') |
                (df_multi_choice['Q6'] == 'Data Engineer') |
                #(df_multi_choice['Q6'] == 'Student') |
                (df_multi_choice['Q6'] == 'Software Engineer')
                ]

df_multi_tech.head()

Time from Start to Finish (seconds)      Q1 Q1_OTHER_TEXT     Q2  \
3                                  718  Female            -1  30-34   
6                                 1142    Male            -1  25-29   
10                                 751    Male            -1  30-34   
12                                 823    Male            -1  40-44   
16                                1153    Male            -1  25-29   

                          Q3                 Q4  \
3   United States of America    Master’s degree   
6                   Colombia  Bachelor’s degree   
10                   Hungary    Master’s degree   
12  United States of America    Master’s degree   
16                     Japan    Master’s degree   

                                               Q5                 Q6  \
3   Computer science (software engineering, etc.)     Data Scientist   
6                            Physics or astronomy     Data Scientist   
10             Engineering (non-computer focused)  Software Engineer   
12             Engineering (non-computer focused)     Data Scientist   
16             Engineering (non-computer focused)     Data Scientist   

   Q6_OTHER_TEXT                                      Q7      ...        \
3             -1                          I am a student      ...         
6             -1                    Computers/Technology      ...         
10            -1  Online Service/Internet-based Services      ...         
12            -1                                   Other      ...         
16            -1                  Medical/Pharmaceutical      ...         

   Q49_OTHER_TEXT Q50_Part_1          Q50_Part_2  \
3              -1        NaN  Too time-consuming   
6              -1        NaN                 NaN   
10             -1        NaN  Too time-consuming   
12             -1        NaN  Too time-consuming   
16             -1        NaN  Too time-consuming   

                               Q50_Part_3  \
3                                     NaN   
6                                     NaN   
10                                    NaN   
12                                    NaN   
16  Requires too much technical knowledge   

                                           Q50_Part_4  \
3                                                 NaN   
6   Afraid that others will use my work without gi...   
10  Afraid that others will use my work without gi...   
12                                                NaN   
16  Afraid that others will use my work without gi...   

                                Q50_Part_5  \
3                                      NaN   
6                                      NaN   
10                                     NaN   
12  Not enough incentives to share my work   
16                                     NaN   

                                           Q50_Part_6 Q50_Part_7 Q50_Part_8  \
3                                                 NaN        NaN        NaN   
6   I had never considered making my work easier f...        NaN        NaN   
10                                                NaN        NaN        NaN   
12                                                NaN        NaN        NaN   
16                                                NaN        NaN        NaN   

   Q50_OTHER_TEXT  
3              -1  
6              -1  
10             -1  
12             -1  
16             -1  

[5 rows x 395 columns]

In [52]:
percent_tech_respondents = (df_multi_tech.shape[0]/df_multi_choice.shape[0])*100 

percent_tech_respondents

33.547089148748896

Note: Data scientists, data engineers, and software engineers account for 33.54% of the total number survey respondents. 

## What is the education breakdown for our group of select respondents? ##

In [98]:
df_test = df_multi_tech.groupby(['Q6','Q4']).size().reset_index(name = 'Percentage')
a = df_test.groupby('Q6')['Percentage'].transform('sum')
df_test['Percentage'] = df_test['Percentage'].div(a).multiply(100)
df_test = df_test.sort_values(by = ['Q6','Percentage'], ascending = False)
df_test.head()

Q6                                                 Q4  \
17  Software Engineer                                    Master’s degree   
14  Software Engineer                                  Bachelor’s degree   
15  Software Engineer                                    Doctoral degree   
20  Software Engineer  Some college/university study without earning ...   
19  Software Engineer                                Professional degree   

    Percentage  
17   43.769968  
14   39.488818  
15    5.910543  
20    4.376997  
19    2.747604

In [100]:
fig = px.bar(df_test, x = 'Q6', y = 'Percentage', title= 'Age and Role of Survey Respondents in 2018', 
             color = 'Q4' , barmode='group', labels={'Percentage':'Percentage (%)'}) #facet_row='Q1')
fig.show()

Key Takeaways: 
    1. A Master's degree is the most common level of education for  the three selected roles. 43.7% of Software Engineers (SWE), 55.4% of Data Scientists, and 55.76% of Data Engineers have a Master's degree. 
    2. 39.49% of Software Engineers (SWE), 20.18% of Data Scientists, and 29.58% of Data Engineers have a Bachelor's degree. A Master's or a Bachelor's degree is more popular than a Doctorate degree. 
    3. Data engineers have the highest percentage of Master's and Bachelor degree graduates. Data Scientists have the highest percentage of doctorage degrees. 

In [136]:
# Salary Breakdown Per Role Per Level of Education

df_test = df_multi_tech.groupby(['Q6', 'Q9', 'Q4']).size().reset_index(name = 'Percentage')
a = df_test.groupby('Q9')['Percentage'].transform('sum')
df_test['Percentage'] = df_test['Percentage'].div(a).multiply(100)
df_test = df_test.sort_values(by = ['Percentage', 'Q6'], ascending = False)

In [144]:
fig = px.bar(df_test, x = 'Q9', y = 'Percentage', title= 'International Compensation Breakdown by Role and Level of Education in 2018', 
             color = 'Q6' , barmode='group', labels={'Percentage':'Percentage (%)'}, facet_row='Q4', height = 3000)
fig.show()

Notes:
    1. The most common amount international compensation for data scientists, SWEs, and Data Engineers with a Master's degree is between 40-50,000, 20-30,000, and 300-400,000, respectively. 

## How does compensation per role and per level of education vary in the Top-3 Most Represented Countries? 

In [197]:
df_test = df_multi_tech.groupby(['Q6', 'Q9', 'Q4', 'Q3']).size().reset_index(name = 'Percentage')
a = df_test.groupby('Q9')['Percentage'].transform('sum')
df_test['Percentage'] = df_test['Percentage'].div(a).multiply(100)

n = 3
df_placeholder = pd.DataFrame({'count' : df_multi_choice.groupby([ "Q3"] ).size()}).reset_index().sort_values(by = 'count', ascending = False)

top_n_countries = df_placeholder['Q3'].unique()[:n]
top_n_countries

df_test = df_test[df_test['Q3'].isin(top_n_countries)]
df_test = df_test.sort_values(by = ['Percentage'], ascending = False)

In [199]:
fig = px.bar(df_test, x = 'Q9', y = 'Percentage', title= ('Top {} Countries: Compensation Breakdown by Role and Level of Education in 2018').format(n), 
             color = 'Q6' , barmode='group', labels={'Percentage':'Percentage (%)', 'Q9':''}, facet_row='Q4', facet_col = 'Q3', height = 3000)
fig.show()

Notes:
Disclaimer: These results do not segment by the years of experience of a respondent. The percentages are grouped by compensation. 

    1. In the United States, the most common compensation for data scientists with a Master's, Doctoral, or Bachelor's degree is 125-150,000 (28.42%), 200-25000+ (21.79%), and 250-300,000 (12.9%), respectively. In India, the most common compensation for data scientists with a Master's, Doctoral, or Bachelor's degree is 400-500,000 (9.09%), 500,000+ (4.17%), and 500,000+ (11.76%), respectively. In China, the most common compensation for data scientists with a Master's, Doctoral, or Bachelor's degree is 500,000+ (5.88%), 500,000+ (5.88%), and 90-100,000 (0.35%), respectively. Key takeaway: Data Scientist wages in the USA are much different in magnitude than in China and India. As a Data Scientist with advanced education -- with a Master's or Doctoral degree -- in China and India, you are most likely to earn 400,000+. A larger percentage of Data Scientists in the USA earn more than 125,000 than in China and India. Although India and China's most common level of compensation for advanced degree holders is greater than in the USA, there are more jobs in the USA for Data Scientists to earn 125,000+. 
    2. Software Engineer Compensation in Top-3 Countries with Varying Levels of Education: In the United States, the most common compensation for data scientists with a Master's, Doctoral, or Bachelor's degree is 300-400,000 (16.67%), 400-500,000+ (18.18%), and 250-300,000+ (12.90%), respectively. In India, the most common compensation for data scientists with a Master's, Doctoral, or Bachelor's degree is 0-10,000 (6.14%), 20-30,000+ (0.34%), and 0-10,000 (14.36%), respectively. In China, the most common compensation for data scientists with a Master's, Doctoral, or Bachelor's degree is 30-40,000 (2.05%), 40-50,000+ (0.22%), and 30-40,000 (1.23%), respectively. Key takeaway: Software engineering wages in the USA far outweigh China and India. Regardless of education level, the most common compensation level in India and China is within a range of 20,000 from a base of 10,000. In the USA, SWEs with doctoral or Master's level education is more likely to earn greater than 100,000 than SWEs with a bachelor's degree.
    3. Data Engineer Compensation in Top-3 Countries with Varying Levels of Education: In the United States, the most common compensation for data engineers with a Master's, Doctoral, or Bachelor's degree is 125-150,000 (4.79%), 200-250,000 (1.28%), and 150-200,000+ (3.48%), respectively. In India, the most common compensation for data engineers with a Master's, Doctoral, or Bachelor's degree is 0-10,000 (1.29%), 40-50,000 (0.22%), and 10-20,000 (1.57%), respectively. In China, the most common compensation for data engineers with a Master's, Doctoral, or Bachelor's degree is 0-10,000 (1.21%), Not disclosed (0.30%), and 10-20,000 (1.29%), respectively. 

## How does compensation and education per role vary in the Top-10 Most Represented Countries? 


## Is it harder for females to break into these roles related to data science? (i.e., do they generally have more education or any other indicators that prove this?)

## What are the barriers data scientists face to scale their work? What skills do they need?


## How are data scientists handling bias in algorithms?

### What is the education level of Data Science and ML respondents? ##